docker/build.sh

```diff
@@ -23,21 +28,67 @@ then
-                                                                 "  git submodule sync --recursive && \n" \
-                                                                 "  git submodule update --init --recursive && \n"
-git am ../../docker/0001-An-approximate-manylinux3.patch
-PLATFORM=$(uname -m) TRAVIS_COMMIT=latest ./build.sh
-popd
+# build manylinux3 if needed
+if [[ "$(docker images -q manylinux3_x86_64 2> /dev/null)" == "" || "$REBUILD_MANYLINUX" != "NO" ]]; then
+    pushd ../third_party/manylinux/
+    git checkout 96b47a25673b33c728e49099a3a6b1bf503a18c2 || echo -e "Did you forget to \`git clone --recursive\`? Try this:\n" \
+                                                                    "  git submodule sync --recursive && \n" \
+                                                                    "  git submodule update --init --recursive && \n"
+    git am ../../docker/0001-An-approximate-manylinux3.patch
+    PLATFORM=$(uname -m) TRAVIS_COMMIT=latest ./build.sh
+    popd
+fi
 
 pushd ../
-docker build -t ${DEPS_IMAGE} --build-arg "FROM_IMAGE_NAME"=manylinux3_x86_64 --build-arg "USE_CUDA_VERSION=${CUDA_VERSION}" -f Dockerfile.deps .
-echo "Build image:" ${BUILDER}
-docker build -t ${BUILDER} --build-arg "DEPS_IMAGE_NAME=${DEPS_IMAGE}" --build-arg "PYVER=${PYVER}" --build-arg "PYV=${PYV}" --build-arg "NVIDIA_BUILD_ID=${NVIDIA_BUILD_ID}" \
+# build deps image if needed
+if [[ "$(docker images -q ${DEPS_IMAGE} 2> /dev/null)" == "" || "$REBUILD_BUILDERS" != "NO" ]]; then
+    echo "Build deps: " ${DEPS_IMAGE}
+    docker build -t ${DEPS_IMAGE} --build-arg "FROM_IMAGE_NAME"=manylinux3_x86_64 --build-arg "USE_CUDA_VERSION=${CUDA_VERSION}" -f Dockerfile.deps .
+fi
+
+# build builder image if needed
+if [[ "$(docker images -q ${BUILDER} 2> /dev/null)" == "" || "$REBUILD_BUILDERS" != "NO" ]]; then
+    echo "Build light image:" ${BUILDER}
+    docker build -t ${BUILDER} --build-arg "DEPS_IMAGE_NAME=${DEPS_IMAGE}" --build-arg "PYVER=${PYVER}" --build-arg "PYV=${PYV}" --build-arg "NVIDIA_BUILD_ID=${NVIDIA_BUILD_ID}" \
                            --build-arg "NVIDIA_DALI_BUILD_FLAVOR=${DALI_BUILD_FLAVOR}" --build-arg "GIT_SHA=${GIT_SHA}" --build-arg "DALI_TIMESTAMP=${DALI_TIMESTAMP}" \
-                           --build-arg "CMAKE_BUILD_TYPE=${BUILD_TYPE}" .
+                           --build-arg "CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" --target builder .
+fi
+
+if [ "$BUILD_INHOST" = "YES" ]; then
+    # build inside the source tree
+    docker run --rm -u 1000:1000 -v $(pwd):/opt/dali ${BUILDER} /bin/bash -c "mkdir -p /opt/dali/${DALI_BUILD_DIR} && \
+                                        cd /opt/dali/${DALI_BUILD_DIR} && \
+                                        rm -rf /opt/dali/${DALI_BUILD_DIR}/nvidia* && \
+                                        CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE:-Release} \
+                                        BUILD_TEST=${BUILD_TEST:-ON}                  \
```

Comment:
Is this a third place where we place defaults into those variables?

Reply:
Done